cmd/log/src/lib.rs

```diff
@@ -89,44 +89,33 @@ fn stringbuf_read(
     core: &mut dyn Core,
     definition: &HubrisStruct,
     ringbuf_var: &HubrisVariable,
-    prev_last_idx: u64,
-) -> Result<(String, u64)> {
+    prev_last_idx: usize,
+) -> Result<(String, usize)> {
     let mut log_msg: String = "".to_owned();
 
     // load the stringbuf from hubris into a corresponding local struct
-    let stringbuf: Stringbuf =
+    let mut stringbuf: Stringbuf =
         load_stringbuf(hubris, core, definition, ringbuf_var);
 
     // extract the log itself as a [u8]
-    let buffer = stringbuf.buffer.as_slice();
+    let buffer = stringbuf.buffer.as_mut_slice();
     // the last written element in the log
-    let last: u64 = stringbuf.last.unwrap() as u64;
+    let last = stringbuf.last.unwrap() as usize;
 
-    // read everything between current last, and last last by
-    // generating index range which connects these two in a circular manner, only including the
-    // new characters, and in the order they should be printed
-    // prev_last_idx has been printed already, last has not
-    let new_idxs: Vec<u64> = if last > prev_last_idx {
-        (prev_last_idx + 1..last).collect()
-    } else {
-        // This logic also works for last == prev_last_idx since we just need to print the entire range
+    let start_read_idx = (prev_last_idx + 1) % buffer.len();
+    // now circular rotate our buffer so it starts with the new characters
+    buffer.rotate_left(start_read_idx as usize);
 
-        // if prev_last_idx was not the last element of the buffer, then we include everything at
-        // the end of the ring buffer
-        if prev_last_idx as usize != buffer.len() - 1 {
-            // should not overflow the length of buffer as we just checked
-            let mut range = ((prev_last_idx + 1) as u64..buffer.len() as u64)
-                .collect::<Vec<u64>>();
-            range.append(&mut (0..=last).collect::<Vec<u64>>());
-            range
-        } else {
-            // prev_last_idx was the last element so we only need the beginning of the buffer
-            (0..=last).collect()
-        }
-    };
+    //
+    // rotate the last index to match the buffer
+    //
+    // we add buffer.len to prevent overflow, it will be "removed" when
+    // we modulus
+    let rot_last = (last + buffer.len()) - start_read_idx;
+    // take the modulus
+    let rot_last = ((rot_last % buffer.len()) + buffer.len()) % buffer.len();
```

Comment:
Shouldn't this be:

```
rot_last = ((last + buffer.len()) - start_read_idx) % buffer.len();
```

The `+ buffer.len()) % buffer.len();` that you have at the end of the second assignment is redundant.

Reply:
Good catch, I forget I already added `buffer.len()`.